File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not  height:
            return 0

        l = 0
        res = 0
        r = len(height) - 1
        lmax, rmax = height[l], height[r]

        while l < r :
            if lmax <= rmax :
                l += 1
                if lmax - height[l] > 0:
                    res += (lmax - height[l])
                lmax = max( lmax, height[l])

            else:
                r -= 1
                if rmax - height[r] > 0:
                    res += ( rmax - height[r])
                rmax = max(rmax, height[r])
        
        return res

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] min = new int[n];
        int[] max = new int[n];

        min[0] = prices[0];
        max[n-1] = prices[n-1];
        for( int i = 1 ; i < n ; i++){
            min[i] = Math.min( min[i-1], prices[i-1]);
        }
        for( int j = n-2; j >= 0 ; j--){
            max[j] = Math.max( max[j+1], prices[j]);
        }
        int res = 0;
        for( int i = 0; i < n; i++){
            res = Math.max(res, max[i] - min[i]);
        }
        return res;
    }
}
